dotrix_core/src/reloadable.rs

```diff
@@ -0,0 +1,90 @@
+//! When an asset changes it may need to reload
+//! and rebind to the pipeline. This Trait is used
+//! for signalling such a reload.
+//!
+
+use std::time::Instant;
+
+#[derive(Debug, Clone, Copy)]
+/// Denotes the kinds of reload/changes of data that can occur
+pub enum ReloadKind {
+    /// No change occured in the last load
+    NoChange,
+    /// Data was updated but a rebind was not required
+    Update,
+    /// Data was updated in a way that requires a rebind
+    Reload,
+}
+
+#[derive(Debug)]
+/// Used to track the state of reloads
+pub struct ReloadState {
+    /// The last instant that an update occured
+    ///
+    /// This is used to signal that update is required
+    /// in assets that are not processed every frame
+    pub last_update_at: Instant,
+    /// The last instant that an reload occured
+    ///
+    /// This is used to signal that reload is required
+    /// in assets that are not processed every frame
+    pub last_reload_at: Instant,
+}
+
+impl Default for ReloadState {
+    fn default() -> Self {
+        Self {
+            last_update_at: Instant::now(),
+            last_reload_at: Instant::now(),
+        }
+    }
+}
+
+/// Describes an object that has gpu represntable data
+/// that may update and needs periodic (re-)loading
+///
+/// A function that needs to for check changes since an Instance
+/// should call [`changed_since`]
+pub trait Reloadable {
```

Comment:
[question]
Could you please describe a use case when there will be just one `Reloadable` or `ReloadableMut` trait implemented?

Reply:
Oh, well when you are rendering you only have access to `&` reference so you can only read the reload state. When you are editing you have an `&mut` reference so you can flag it as dirty. You cannot have a function that takes an `&mut self` in a trait impl of `& MyStruct` so you have to seperate them.

Reply:
I will see if I can explain this.

At bind you have:

```rust
Binding::Texture("label", Stage::All, Box<& Texture>(asset)
```

Notice that it is a `&Texture` which comes from `assets.get<Texture>(id)`.

The trait `Reloadable` cannot have mut assess to this, so the trait impl must not have any `&mut self`  functions if I want to use this trait.